### PR TITLE
sys-libs/llvm-libunwind: new package

### PR DESCRIPTION
**Pull request by zzlei**
This is LLVM's alternative implementation of libunwind. libcxxrt is also revised to support llvm-libunwind.

### CODE REVIEW

File: sys-libs/libcxxrt/libcxxrt-9999.ebuild [hunk 1]
```diff
@@ -28,7 +28,8 @@ else
 fi
 IUSE="libunwind +static-libs"
 
-RDEPEND="libunwind? ( >=sys-libs/libunwind-1.0.1-r1[static-libs?] )"
+RDEPEND="libunwind? ( || ( >=sys-libs/libunwind-1.0.1-r1[static-libs?] \
+	sys-libs/llvm-libunwind[static-libs?] ) )"
```
Comment (lu-zero) [inline sys-libs/libcxxrt/libcxxrt-9999.ebuild]: I think it is ok for now.

File: sys-libs/libcxxrt/libcxxrt-0.0_p20150423-r1.ebuild [hunk 2]
```diff
@@ -28,7 +28,8 @@ else
 fi
 IUSE="libunwind +static-libs"
 
-RDEPEND="libunwind? ( >=sys-libs/libunwind-1.0.1-r1[static-libs?] )"
+RDEPEND="libunwind? ( || ( >=sys-libs/libunwind-1.0.1-r1[static-libs?] \
```
Comment (mgorny) [inline sys-libs/libcxxrt/libcxxrt-0.0_p20150423-r1.ebuild]: The backslash here is unnecessary.

File: sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Copyright 1999-2016 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Id$
+
+EAPI=5
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]: Why not EAPI=6?

File: sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+DESCRIPTION="C++ runtime stack unwinder from LLVM"
+HOMEPAGE="https://github.com/llvm-mirror/libunwind"
+if [ "${PV%9999}" = "${PV}" ] ; then
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]: `[[ ${PV} == 9999 ]]`

File: sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+LICENSE="|| ( UoI-NCSA MIT )"
+SLOT="0"
+IUSE="+static-libs"
+
+DEPEND=""
+RDEPEND="!sys-libs/libunwind"
+
+src_unpack() {
+	default
+	S="${WORKDIR}/libunwind-${PV}.src"
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]: The standard practice is to set it in global scope. It'd also make sense to define MY_P and use it to form both SRC_URI and S.

File: sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild (hunk 4 above)
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]: Not that this really makes sense considering the live ebuild stuff is not supported anywhere else.

File: sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+src_prepare() {
+	cp "${FILESDIR}/Makefile" src/ || die
+}
+
+src_compile() {
+	cd src || die
+	emake shared
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]:
```
emake -C src ...

File: sys-libs/llvm-libunwind/llvm-libunwind-9999.ebuild [hunk 7]
```diff
@@ -0,0 +1,42 @@
+# Copyright 1999-2016 Gentoo Foundation
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-9999.ebuild]: Why are you adding a half-written, non-working ebuild?

File: sys-libs/llvm-libunwind/metadata.xml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+	<maintainer type="person">
+		<email>zhanglei.april@gmail.com</email>
+		<name>Lei Zhang</name>
+	</maintainer>
+</pkgmetadata>
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/metadata.xml]: Add also proxy-maint here.

File: sys-libs/llvm-libunwind/files/Makefile [hunk 9]
```diff
@@ -0,0 +1,43 @@
+ASFLAGS+=-Wa,--noexecstack
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/files/Makefile]: What's the reason you're reinventing the build system instead of using upstream's?

Comment (zzlei) [inline sys-libs/llvm-libunwind/files/Makefile]: Its native build system is cmake, but cmake needs to know the path to llvm's source to configure correctly, although libunwind doesn't depend on any file from llvm.

libcxx also faces the same problem. I personally hate to do it this way, and will see I can fix this cmake nonsense...

File: sys-libs/llvm-libunwind/metadata.xml (hunk 8 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/metadata.xml]: > Add also proxy-maint here.

I'm not sure what this means. Should I list a Gentoo dev here? @lu-zero

Comment (mgorny) [inline sys-libs/llvm-libunwind/metadata.xml]: No, the proxy-maint project. The people who will review your PRs and stuff, and push them on your behalf. Grep existing metadata.xml files for example on how it should look like.

File: sys-libs/llvm-libunwind/files/Makefile (hunk 9 above)
Comment (mgorny) [inline sys-libs/llvm-libunwind/files/Makefile]: Please try. LLVM upstream is quite likely to accept a good patch to support this.

File: sys-libs/llvm-libunwind/llvm-libunwind-9999.ebuild (hunk 7 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/llvm-libunwind-9999.ebuild]: Thanks for all the corrections :)

File: sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild (hunk 4 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]: I thought ${PN}-9999.ebuild serves as kind of a template ebuild for its package. No?

Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-3.8.1.ebuild]: Not at all. Normally it's a live ebuild building stuff from git/svn.

File: sys-libs/llvm-libunwind/metadata.xml (hunk 8 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/metadata.xml]: Before I list a proxy-maint project here, do I need to ask for their permission?

Comment (mgorny) [inline sys-libs/llvm-libunwind/metadata.xml]: You have my blessing.

File: sys-libs/llvm-libunwind/files/Makefile (hunk 9 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/files/Makefile]: Now I dropped my Makefile and switched to cmake, with a downstream patch. This patch needs more work to be upstreamed though.

File: sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+index ee01e9a..c66673e 100644
+--- a/src/CMakeLists.txt
++++ b/src/CMakeLists.txt
+@@ -18,7 +18,8 @@ set(LIBUNWIND_ASM_SOURCES
+     UnwindRegistersSave.S)
+ set_source_files_properties(${LIBUNWIND_ASM_SOURCES}
+                             PROPERTIES
+-                              LANGUAGE C)
++                              LANGUAGE C
++                              COMPILE_FLAGS "-Wa,--noexecstack")
```
Comment (lu-zero) [inline sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch]: The cmake eclass doesn't take care of it already?

File: sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+@@ -51,8 +52,10 @@ set(LIBUNWIND_SOURCES
+ 
+ if (LIBUNWIND_ENABLE_SHARED)
+   add_library(unwind SHARED ${LIBUNWIND_SOURCES} ${LIBUNWIND_HEADERS})
+-else()
+-  add_library(unwind STATIC ${LIBUNWIND_SOURCES} ${LIBUNWIND_HEADERS})
++endif ()
++if (LIBUNWIND_ENABLE_STATIC)
++  add_library(unwind_static STATIC ${LIBUNWIND_SOURCES} ${LIBUNWIND_HEADERS})
++  set_target_properties(unwind_static PROPERTIES OUTPUT_NAME unwind)
```
Comment (lu-zero) [inline sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch]: You might discuss shared/static support with upstream I guess.

File: sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch (hunk 10 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch]: Without setting this flag manually, I get warning from emerge about executable stack. I suppose they're handled in llvm's cmake scripts, which aren't included in the standalone build.

**Comment by zzlei** [comment]
I'm closing this PR to combine it with my other related PRs into a new one, for the sake of easier reviewing.

File: sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch (hunk 11 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch]: will do :)

File: sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch (hunk 10 above)
Comment (zzlei) [inline sys-libs/llvm-libunwind/files/libunwind-3.8-cmake.patch]: Just checked. llvm's cmake scripts doesn't take care of it either.

**Pull request closed without merging**